File: 2226 maximum-candies-allocated-to-k-children/solution.cpp
class Solution {
public:
    int maximumCandies(vector<int>& candies, long long k) {
        int l = 1,r = 0, ans = 0;
        for(int i:candies) r=max(r,i);
        while(l<=r) {
            int mid = l+(r-l)/2;
            if(pos(candies,mid,k)) {
                l=mid+1;
                ans=mid;
            } else {
                r=mid-1;
            }            
        }
        return ans;
    }
    bool pos(vector<int> nums,int n,long long k) {
        long long c = 0;
        for(int i:nums) {
            c+=i/n;
        }
        return c>=k;
    }
};

File: 2226 maximum-candies-allocated-to-k-children/solution.rs
impl Solution {
    pub fn maximum_candies(candies: Vec<i32>, k: i64) -> i32 {
        let mut l = 1;
        let mut r = *candies.iter().max().unwrap();
        let mut ans = 0;

        while l <= r {
            let mid = l + (r - l) / 2;
            if Self::pos(&candies, mid, k) {
                ans = mid;
                l = mid + 1;
            }   else {
                r = mid - 1;
            }
        }
        ans
    }

    fn pos(nums: &Vec<i32>, n: i32, k: i64) -> bool {
        let mut c: i64 = 0;
        for &v in nums {
            c += (v / n) as i64;
        }
        c >= k
    }
}
